lz4/stream/_stream.c

```diff
@@ -0,0 +1,1404 @@
+#define LZ4_VERSION_GE_1_9_0  1
+#pragma message "Building bindings against LZ4 " LZ4_VERSION_STRING ", using API>=1.9.0: yes"
+#else
+#define LZ4_VERSION_GE_1_9_0  0
+#pragma message "Building bindings against LZ4 " LZ4_VERSION_STRING ", using API>=1.9.0: no"
+#endif
+
+#ifndef Py_UNUSED /* This is already defined for Python 3.4 onwards */
+#ifdef __GNUC__
+#define Py_UNUSED(name) _unused_ ## name __attribute__((unused))
+#else
+#define Py_UNUSED(name) _unused_ ## name
+#endif
+#endif
+
+#if defined(_WIN32) && defined(_MSC_VER)
+#if _MSC_VER >= 1600
+#include <stdint.h>
+#else /* _MSC_VER >= 1600 */
+typedef signed char int8_t;
+typedef signed short int16_t;
+typedef signed int int32_t;
+typedef unsigned char uint8_t;
+typedef unsigned short uint16_t;
+typedef unsigned int uint32_t;
+#endif /* _MSC_VER >= 1600 */
+#if !defined(__CHAR_BIT__)
+#define __CHAR_BIT__ 8
+#endif /* __CHAR_BIT__ */
+#endif
+
+
+typedef enum
+{
+  COMPRESS,
+  DECOMPRESS,
+} lz4_direction_e;
+
+typedef enum
+{
+  DEFAULT,
+  FAST,
+  HIGH_COMPRESSION
+} compression_type;
+
+
+/* Forward declarations */
+static PyObject * LZ4StreamError;
+static PyTypeObject _LZ4StreamContext;
+
+#define DOUBLE_BUFFER_PAGE_COUNT          (2)
+
+#define DOUBLE_BUFFER_INDEX_MIN           (0)
+#define DOUBLE_BUFFER_INDEX_INVALID       (-1)
+
+#define _GET_MAX_UINT(byte_depth, type)   (type)( ( 1ULL << ( __CHAR_BIT__ * (byte_depth) ) ) - 1 )
+#define _GET_MAX_UINT32(byte_depth)       _GET_MAX_UINT((byte_depth), uint32_t)
+
+typedef struct {
+  char         * buf;
+  unsigned int   len;
+} buffer_t;
+
+typedef struct {
+  PyObject_HEAD
+  /* Type-specific fields go here. */
+
+  /* Double buffer */
+  struct {
+    char         * buf;
+    unsigned int   page_size;
+    char         * pages[DOUBLE_BUFFER_PAGE_COUNT];
+    int            index;
+  } double_buffer;
+
+  buffer_t output;
+
+  /* LZ4 state */
+  union {
+    union {
+      LZ4_stream_t   * fast;
+      LZ4_streamHC_t * hc;
+    } compress;
+    LZ4_streamDecode_t * decompress;
+  } lz4_state;
+
+  /* LZ4 configuration */
+  struct {
+    int              acceleration;
+    int              compression_level;
+    int              store_comp_size;
+    int              return_bytearray;
+    lz4_direction_e  direction;
+    compression_type comp;
+  } config;
+} _LZ4StreamContext_t;
+
+static inline void
+store_le8 (char * c, uint8_t x)
+{
+  c[0] = x & 0xff;
+}
+
+static inline uint8_t
+load_le8 (const char * c)
+{
+  const uint8_t * d = (const uint8_t *) c;
+  return d[0];
+}
+
+static inline void
+store_le16 (char * c, uint16_t x)
+{
+  c[0] = x & 0xff;
+  c[1] = (x >> 8) & 0xff;
+}
+
+static inline uint16_t
+load_le16 (const char * c)
+{
+  const uint8_t * d = (const uint8_t *) c;
+  return (d[0] | (d[1] << 8));
+}
+
+static inline void
+store_le32 (char * c, uint32_t x)
+{
+  c[0] = x & 0xff;
+  c[1] = (x >> 8) & 0xff;
+  c[2] = (x >> 16) & 0xff;
+  c[3] = (x >> 24) & 0xff;
+}
+
```

Comment:
As above - how useful in practice is it to be able to store the size as 1, 2, or 4 bytes? I am happy to keep it if there's a geneuine use case, but I tend to favour simlicity wherever possible.

Reply:
@jonathanunderwood Actually my use case needs this (the size of the space reserved to write the compressed data length) being configurable - I handle streamed compressed data in which the compressed block size is on 2 bytes and this cannot be changed.

This is the reason I introduced the `store_comp_size` parameter in the context, which its default value is `4`.